File: typescript/tests/compute-catalog-integrity.test.ts
/**
 * Compute catalog integrity — structure, Decimal parsing, freshness,
 * dispatch coverage.
 *
 * Ports python/tests/test_compute_catalog_integrity.py (13 cases) to vitest.
 */

import { describe, expect, test } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import Decimal from "decimal.js";

const HERE = dirname(fileURLToPath(import.meta.url));
const CATALOG_PATH = join(HERE, "../src/data/compute_prices.json");

function load(): Record<string, any> {
  return JSON.parse(readFileSync(CATALOG_PATH, "utf-8")) as Record<string, any>;
}

describe("compute catalog integrity", () => {
  test("catalog parses as JSON with _meta", () => {
    const data = load();
    expect(data).toHaveProperty("_meta");
  });

  test("_meta has required default keys", () => {
    const data = load();
    const meta = data._meta as Record<string, any>;
    const required = [
      "version",
      "last_updated",
      "currency",
      "default_lambda_request_usd",
      "default_lambda_gb_second_usd",
      "default_fargate_vcpu_second_usd",
      "default_fargate_gib_second_usd",
      "default_cloud_run_request_usd",
      "default_cloud_run_vcpu_second_usd",
      "default_cloud_run_gib_second_usd",
      "default_azure_functions_execution_usd",
      "default_azure_functions_gb_second_usd",
      "default_vercel_cpu_hour_usd",
      "default_vercel_memory_gb_hour_usd",
      "default_ec2_vcpu_hour_usd",
      "default_k8s_pod_vcpu_hour_usd",
      "description",
      "notes",
    ];
    for (const k of required) {
      expect(meta, `_meta missing ${k}`).toHaveProperty(k);
      if (k.startsWith("default_") && k.endsWith("_usd")) {
        // Must parse as Decimal.
        expect(() => new Decimal(meta[k])).not.toThrow();
      }
    }
    expect(meta.currency).toBe("USD");
  });

  test("every provider has _last_verified", () => {
    const data = load();
    const today = new Date();
    const softLimitDays = 180;
    for (const [provider, block] of Object.entries(data)) {
      if (provider === "_meta") continue;
      const verifiedRaw = (block as any)._last_verified;
      expect(verifiedRaw, `${provider} missing _last_verified`).toBeDefined();
      const verified = new Date(verifiedRaw);
      const days = Math.floor(
        (today.getTime() - verified.getTime()) / (24 * 3600 * 1000),
      );
      if (days > softLimitDays) {
        // Soft warn — do not fail.
        // eslint-disable-next-line no-console
        console.warn(
          `compute_prices.json: ${provider} _last_verified is ${days} days old (soft limit ${softLimitDays})`,
        );
      }
    }
  });

  test("all providers and runtimes present", () => {
    const data = load();
    for (const p of ["aws", "gcp", "azure", "vercel"]) {
      expect(data).toHaveProperty(p);
    }
    const awsRuntimes = new Set(Object.keys(data.aws).filter((k) => k !== "_last_verified"));
    for (const r of ["lambda", "fargate", "ec2"]) {
      expect(awsRuntimes.has(r), `aws missing runtime ${r}`).toBe(true);
    }
    const gcpRuntimes = new Set(Object.keys(data.gcp).filter((k) => k !== "_last_verified"));
    for (const r of ["cloud_run", "cloud_functions", "gce"]) {
      expect(gcpRuntimes.has(r), `gcp missing runtime ${r}`).toBe(true);
    }
    const azureRuntimes = new Set(Object.keys(data.azure).filter((k) => k !== "_last_verified"));
    for (const r of ["functions_consumption", "vm"]) {
      expect(azureRuntimes.has(r), `azure missing runtime ${r}`).toBe(true);
    }
    expect(data.vercel).toHaveProperty("fluid");
  });

  test("Lambda has both architectures", () => {
    const data = load();
    const def = data.aws.lambda.default;
    expect(new Set(Object.keys(def))).toEqual(new Set(["x86_64", "arm64"]));
    for (const arch of ["x86_64", "arm64"]) {
      expect(() => new Decimal(def[arch].request_usd)).not.toThrow();
      expect(() => new Decimal(def[arch].gb_second_usd)).not.toThrow();
    }
  });

  test("Fargate has both architectures", () => {
    const data = load();
    const def = data.aws.fargate.default;
    expect(new Set(Object.keys(def))).toEqual(new Set(["x86_64", "arm64"]));
    for (const arch of ["x86_64", "arm64"]) {
      expect(() => new Decimal(def[arch].vcpu_second_usd)).not.toThrow();
      expect(() => new Decimal(def[arch].gib_second_usd)).not.toThrow();
    }
  });

  test("ARM cheaper than x86 on Lambda", () => {
    const data = load();
    const firstRegion = Object.values<any>(data.aws.lambda.regions)[0];
    const arm = new Decimal(firstRegion.arm64.gb_second_usd);
    const x86 = new Decimal(firstRegion.x86_64.gb_second_usd);
    expect(arm.lt(x86), "arm64 must be cheaper than x86_64 on Lambda").toBe(true);
  });

  test("ARM cheaper than x86 on Fargate", () => {
    const data = load();
    const firstRegion = Object.values<any>(data.aws.fargate.regions)[0];
    const arm = new Decimal(firstRegion.arm64.vcpu_second_usd);
    const x86 = new Decimal(firstRegion.x86_64.vcpu_second_usd);
    expect(arm.lt(x86), "arm64 must be cheaper than x86_64 on Fargate").toBe(true);
  });

  test("top EC2 SKUs present for us-east-1", () => {
    const data = load();
    const its = data.aws.ec2.regions["us-east-1"].instance_types;
    for (const sku of ["c7g.xlarge", "m7i.large", "t3.medium"]) {
      expect(its, `missing EC2 SKU: ${sku}`).toHaveProperty(sku);
      expect(() => new Decimal(its[sku].hourly_usd)).not.toThrow();
      expect(() => new Decimal(its[sku].vcpu_count)).not.toThrow();
    }
  });

  test("top GCE SKUs present for us-central1", () => {
    const data = load();
    const its = data.gcp.gce.regions["us-central1"].instance_types;
    for (const sku of ["n2-standard-2", "e2-standard-4"]) {
      expect(its, `missing GCE SKU: ${sku}`).toHaveProperty(sku);
    }
  });

  test("top Azure VM SKUs present for eastus", () => {
    const data = load();
    const its = data.azure.vm.regions["eastus"].instance_types;
    for (const sku of ["Standard_D2s_v3", "Standard_B2ms"]) {
      expect(its, `missing Azure VM SKU: ${sku}`).toHaveProperty(sku);
    }
  });

  test("every rate is Decimal-parseable", () => {
    const data = load();

    function walk(node: unknown, path: string): void {
      if (node !== null && typeof node === "object" && !Array.isArray(node)) {
        for (const [k, v] of Object.entries(node as Record<string, unknown>)) {
          walk(v, `${path}.${k}`);
        }
      } else if (typeof node === "string") {
        if (path.endsWith("_usd") || path.endsWith("vcpu_count")) {
          expect(() => new Decimal(node), `${path} not Decimal-parseable: ${node}`).not.toThrow();
        }
      }
    }
    walk(data, "");
  });

  test("every dispatch billing_model has a rate path", () => {
    const data = load();
    const meta = data._meta;
    for (const k of [
      "default_lambda_request_usd",
      "default_fargate_vcpu_second_usd",
      "default_cloud_run_request_usd",
      "default_azure_functions_execution_usd",
      "default_vercel_cpu_hour_usd",
      "default_ec2_vcpu_hour_usd",
      "default_k8s_pod_vcpu_hour_usd",
    ]) {
      expect(meta).toHaveProperty(k);
    }
  });
});

File: go/pricing/compute_catalog_integrity_test.go
// Compute catalog integrity — structure, Decimal parsing, freshness,
// dispatch coverage. Mirrors python/tests/test_compute_catalog_integrity.py.

package pricing

import (
	_ "embed"
	"encoding/json"
	"strings"
	"testing"
	"time"

	"github.com/shopspring/decimal"
)

//go:embed data/compute_prices.json
var embeddedComputeCatalogForTests []byte

func loadComputeCatalog(t *testing.T) map[string]any {
	t.Helper()
	var data map[string]any
	if err := json.Unmarshal(embeddedComputeCatalogForTests, &data); err != nil {
		t.Fatalf("compute_prices.json parse: %v", err)
	}
	return data
}

func TestComputeCatalogParsesAsJSON(t *testing.T) {
	data := loadComputeCatalog(t)
	if _, ok := data["_meta"]; !ok {
		t.Fatal("_meta missing")
	}
}

func TestComputeMetaHasRequiredDefaultKeys(t *testing.T) {
	data := loadComputeCatalog(t)
	meta, ok := data["_meta"].(map[string]any)
	if !ok {
		t.Fatal("_meta not a map")
	}
	required := []string{
		"version", "last_updated", "currency",
		"default_lambda_request_usd", "default_lambda_gb_second_usd",
		"default_fargate_vcpu_second_usd", "default_fargate_gib_second_usd",
		"default_cloud_run_request_usd", "default_cloud_run_vcpu_second_usd",
		"default_cloud_run_gib_second_usd",
		"default_azure_functions_execution_usd",
		"default_azure_functions_gb_second_usd",
		"default_vercel_cpu_hour_usd", "default_vercel_memory_gb_hour_usd",
		"default_ec2_vcpu_hour_usd", "default_k8s_pod_vcpu_hour_usd",
		"description", "notes",
	}
	for _, k := range required {
		v, present := meta[k]
		if !present {
			t.Errorf("_meta missing %s", k)
			continue
		}
		if strings.HasPrefix(k, "default_") && strings.HasSuffix(k, "_usd") {
			s, ok := v.(string)
			if !ok {
				t.Errorf("_meta.%s not a string: %T", k, v)
				continue
			}
			if _, err := decimal.NewFromString(s); err != nil {
				t.Errorf("_meta.%s = %q not Decimal-parseable: %v", k, s, err)
			}
		}
	}
	if meta["currency"] != "USD" {
		t.Errorf("currency = %v, want USD", meta["currency"])
	}
}

func TestComputeEveryProviderHasLastVerified(t *testing.T) {
	data := loadComputeCatalog(t)
	today := time.Now().UTC()
	softLimit := 180 * 24 * time.Hour
	for provider, raw := range data {
		if provider == "_meta" {
			continue
		}
		block, ok := raw.(map[string]any)
		if !ok {
			t.Errorf("provider %s not a map", provider)
			continue
		}
		v, ok := block["_last_verified"].(string)
		if !ok {
			t.Errorf("provider %s missing _last_verified", provider)
			continue
		}
		verified, err := time.Parse("2006-01-02", v)
		if err != nil {
			t.Errorf("provider %s _last_verified %q not ISO date: %v", provider, v, err)
			continue
		}
		if age := today.Sub(verified); age > softLimit {
			// Soft warn — t.Logf (NOT t.Errorf) per plan.
			t.Logf("compute_prices.json: %s _last_verified is %d days old (soft limit 180)",
				provider, int(age.Hours()/24))
		}
	}
}

func TestComputeAllProvidersAndRuntimesPresent(t *testing.T) {
	data := loadComputeCatalog(t)
	for _, p := range []string{"aws", "gcp", "azure", "vercel"} {
		if _, ok := data[p]; !ok {
			t.Errorf("provider %s missing", p)
		}
	}
	aws, _ := data["aws"].(map[string]any)
	for _, r := range []string{"lambda", "fargate", "ec2"} {
		if _, ok := aws[r]; !ok {
			t.Errorf("aws.%s missing", r)
		}
	}
	gcp, _ := data["gcp"].(map[string]any)
	for _, r := range []string{"cloud_run", "cloud_functions", "gce"} {
		if _, ok := gcp[r]; !ok {
			t.Errorf("gcp.%s missing", r)
		}
	}
	azure, _ := data["azure"].(map[string]any)
	for _, r := range []string{"functions_consumption", "vm"} {
		if _, ok := azure[r]; !ok {
			t.Errorf("azure.%s missing", r)
		}
	}
	vercel, _ := data["vercel"].(map[string]any)
	if _, ok := vercel["fluid"]; !ok {
		t.Error("vercel.fluid missing")
	}
}

func TestLambdaHasBothArchitectures(t *testing.T) {
	data := loadComputeCatalog(t)
	def := getPath(t, data, "aws", "lambda", "default").(map[string]any)
	for _, arch := range []string{"x86_64", "arm64"} {
		ad, ok := def[arch].(map[string]any)
		if !ok {
			t.Errorf("aws.lambda.default missing %s", arch)
			continue
		}
		mustDecimal(t, ad, "request_usd")
		mustDecimal(t, ad, "gb_second_usd")
	}
}

func TestFargateHasBothArchitectures(t *testing.T) {
	data := loadComputeCatalog(t)
	def := getPath(t, data, "aws", "fargate", "default").(map[string]any)
	for _, arch := range []string{"x86_64", "arm64"} {
		ad, ok := def[arch].(map[string]any)
		if !ok {
			t.Errorf("aws.fargate.default missing %s", arch)
			continue
		}
		mustDecimal(t, ad, "vcpu_second_usd")
		mustDecimal(t, ad, "gib_second_usd")
	}
}

func TestARMCheaperThanX86OnLambda(t *testing.T) {
	data := loadComputeCatalog(t)
	regions := getPath(t, data, "aws", "lambda", "regions").(map[string]any)
	var any1 map[string]any
	for _, v := range regions {
		any1 = v.(map[string]any)
		break
	}
	arm := decimal.RequireFromString(any1["arm64"].(map[string]any)["gb_second_usd"].(string))
	x86 := decimal.RequireFromString(any1["x86_64"].(map[string]any)["gb_second_usd"].(string))
	if !arm.LessThan(x86) {
		t.Fatalf("arm64 (%s) must be cheaper than x86_64 (%s) on Lambda", arm, x86)
	}
}

func TestARMCheaperThanX86OnFargate(t *testing.T) {
	data := loadComputeCatalog(t)
	regions := getPath(t, data, "aws", "fargate", "regions").(map[string]any)
	var any1 map[string]any
	for _, v := range regions {
		any1 = v.(map[string]any)
		break
	}
	arm := decimal.RequireFromString(any1["arm64"].(map[string]any)["vcpu_second_usd"].(string))
	x86 := decimal.RequireFromString(any1["x86_64"].(map[string]any)["vcpu_second_usd"].(string))
	if !arm.LessThan(x86) {
		t.Fatalf("arm64 (%s) must be cheaper than x86_64 (%s) on Fargate", arm, x86)
	}
}

func TestTopInstanceTypesPresentForEC2USEast1(t *testing.T) {
	data := loadComputeCatalog(t)
	regions := getPath(t, data, "aws", "ec2", "regions").(map[string]any)
	use1, ok := regions["us-east-1"].(map[string]any)
	if !ok {
		t.Fatal("aws.ec2.regions.us-east-1 missing")
	}
	instances, ok := use1["instance_types"].(map[string]any)
	if !ok {
		t.Fatal("us-east-1.instance_types missing")
	}
	for _, want := range []string{"c7g.xlarge", "m7i.large", "t3.medium"} {
		v, ok := instances[want]
		if !ok {
			t.Errorf("missing EC2 SKU: %s", want)
			continue
		}
		row, _ := v.(map[string]any)
		mustDecimal(t, row, "hourly_usd")
		mustDecimal(t, row, "vcpu_count")
	}
}

func TestTopInstanceTypesPresentForGCEUSCentral1(t *testing.T) {
	data := loadComputeCatalog(t)
	instances := getPath(t, data, "gcp", "gce", "regions", "us-central1", "instance_types").(map[string]any)
	for _, want := range []string{"n2-standard-2", "e2-standard-4"} {
		if _, ok := instances[want]; !ok {
			t.Errorf("missing GCE SKU: %s", want)
		}
	}
}

func TestTopInstanceTypesPresentForAzureVMEastus(t *testing.T) {
	data := loadComputeCatalog(t)
	instances := getPath(t, data, "azure", "vm", "regions", "eastus", "instance_types").(map[string]any)
	for _, want := range []string{"Standard_D2s_v3", "Standard_B2ms"} {
		if _, ok := instances[want]; !ok {
			t.Errorf("missing Azure VM SKU: %s", want)
		}
	}
}

func TestEveryComputeRateIsDecimalParseable(t *testing.T) {
	data := loadComputeCatalog(t)
	var walk func(node any, path string)
	walk = func(node any, path string) {
		switch v := node.(type) {
		case map[string]any:
			for k, child := range v {
				walk(child, path+"."+k)
			}
		case string:
			if strings.HasSuffix(path, "_usd") || strings.HasSuffix(path, "vcpu_count") {
				if _, err := decimal.NewFromString(v); err != nil {
					t.Errorf("%s not Decimal-parseable: %q (%v)", path, v, err)
				}
			}
		}
	}
	walk(data, "")
}

func TestEveryDispatchBillingModelHasARatePath(t *testing.T) {
	data := loadComputeCatalog(t)
	meta := data["_meta"].(map[string]any)
	required := []string{
		"default_lambda_request_usd",
		"default_fargate_vcpu_second_usd",
		"default_cloud_run_request_usd",
		"default_azure_functions_execution_usd",
		"default_vercel_cpu_hour_usd",
		"default_ec2_vcpu_hour_usd",
		"default_k8s_pod_vcpu_hour_usd",
	}
	for _, k := range required {
		if _, ok := meta[k]; !ok {
			t.Errorf("_meta.%s missing — billing model unreachable", k)
		}
	}
}

// getPath walks a nested map by string keys, failing the test on miss.
func getPath(t *testing.T, root any, keys ...string) any {
	t.Helper()
	cur := root
	for _, k := range keys {
		m, ok := cur.(map[string]any)
		if !ok {
			t.Fatalf("path %v: not a map at %q", keys, k)
		}
		cur, ok = m[k]
		if !ok {
			t.Fatalf("path %v: missing key %q", keys, k)
		}
	}
	return cur
}

func mustDecimal(t *testing.T, m map[string]any, key string) {
	t.Helper()
	s, ok := m[key].(string)
	if !ok {
		t.Errorf("%s not a string", key)
		return
	}
	if _, err := decimal.NewFromString(s); err != nil {
		t.Errorf("%s = %q not Decimal-parseable: %v", key, s, err)
	}
}
